File: src/main/java/mediathek/gui/filmInformation/InfoDialog.java
package mediathek.gui.filmInformation;

import com.jidesoft.swing.MultilineLabel;
import javafx.application.Platform;
import javafx.embed.swing.JFXPanel;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.Hyperlink;
import javafx.scene.control.MenuItem;
import javafx.scene.control.Tooltip;
import javafx.scene.layout.Background;
import javafx.scene.layout.BackgroundFill;
import javafx.scene.layout.CornerRadii;
import javafx.scene.paint.Color;
import mediathek.config.Daten;
import mediathek.daten.DatenFilm;
import mediathek.gui.actions.UrlHyperlinkAction;
import mediathek.tool.ApplicationConfiguration;
import mediathek.tool.GuiFunktionen;
import mediathek.tool.MVSenderIconCache;
import net.miginfocom.layout.CC;
import net.miginfocom.swing.MigLayout;
import org.apache.commons.configuration2.Configuration;
import org.apache.commons.configuration2.sync.LockMode;

import javax.swing.*;
import javax.swing.text.JTextComponent;
import java.awt.*;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.net.URISyntaxException;
import java.util.NoSuchElementException;
import java.util.Optional;

public class InfoDialog extends JDialog {
    private static final long serialVersionUID = -890508930316467747L;
    private static final String FILM_INFO_VISIBLE = "film.information.visible";
    private static final String FILM_INFO_LOCATION_X = "film.information.location.x";
    private static final String FILM_INFO_LOCATION_Y = "film.information.location.y";
    private static final String COPY_URL_TEXT = "URL kopieren";
    private final Configuration config = ApplicationConfiguration.getConfiguration();
    private final MVSenderIconCache senderIconCache;
    private DatenFilm currentFilm = null;
    private JLabel lblSender;
    private MultilineLabel lblThema;
    private MultilineLabel lblTitel;
    private JLabel lblDatum;
    private JLabel lblUhrzeit;
    private JLabel lblDauer;
    private JLabel lblSize;
    private JCheckBox cbHD;
    private JCheckBox cbSubtitle;
    private JLabel lblGeo;
    private JLabel lblAbo;
    private JFXPanel fxPanel;
    private Hyperlink hyperlink;
    private JTextArea lblDescription;

    public InfoDialog(Window parent) {
        super(parent);
        senderIconCache = Daten.getInstance().getSenderIconCache();

        setType(Window.Type.UTILITY);
        setTitle("Filminformation");
        setResizable(false);
        setDefaultCloseOperation(DISPOSE_ON_CLOSE);

        buildLayout();

        installCopyUrlHandler(lblThema);
        installCopyUrlHandler(lblTitel);

        updateTextFields();

        restoreLocation();
        final boolean wasVisible = config.getBoolean(FILM_INFO_VISIBLE, false);
        if (wasVisible) {
            setVisible(true);
        }

        addWindowListener(new WindowAdapter() {
            @Override
            public void windowOpened(WindowEvent e) {
                config.setProperty(FILM_INFO_VISIBLE, true);
            }

            @Override
            public void windowClosed(WindowEvent e) {
                config.setProperty(FILM_INFO_VISIBLE, false);
            }
        });

        //addFilmlistLoadListener();

        addComponentListener(new ComponentAdapter() {
            @Override
            public void componentMoved(ComponentEvent e) {
                if (isVisible())
                    saveLocation();
            }
        });
    }

    private void installCopyUrlHandler(JTextComponent component) {
        JPopupMenu menu = new JPopupMenu();
        JMenuItem item = new JMenuItem(COPY_URL_TEXT);
        item.addActionListener(e -> GuiFunktionen.copyToClipboard(component.getText()));
        menu.add(item);
        component.setComponentPopupMenu(menu);
    }

    /**
     * Restore window position from config settings.
     */
    private void restoreLocation() {
        config.lock(LockMode.READ);
        try {
            final Point newLocation = new Point();
            newLocation.x = config.getInt(FILM_INFO_LOCATION_X);
            newLocation.y = config.getInt(FILM_INFO_LOCATION_Y);
            setLocation(newLocation);
        } catch (NoSuchElementException ignored) {
        } finally {
            config.unlock(LockMode.READ);
        }
    }

    /**
     * Save window position to config.
     */
    private void saveLocation() {
        //prevent strange OpenJDK bug on Linux where getLocationOnScreen will fail if not visible...
        if (!isVisible())
            return;

        config.lock(LockMode.WRITE);
        try {
            final Point location = getLocationOnScreen();
            config.setProperty(FILM_INFO_LOCATION_X, location.x);
            config.setProperty(FILM_INFO_LOCATION_Y, location.y);
        } finally {
            config.unlock(LockMode.WRITE);
        }
    }

    public void showInfo() {
        updateTextFields();
        pack();

        if (!isVisible())
            setVisible(true);

    }

    private void clearControls() {
        lblSender.setText("");
        lblSender.setIcon(null);
        lblThema.setText("");
        lblTitel.setText("");
        lblDescription.setText("");
        lblSize.setText("");
        lblDatum.setText("");
        lblUhrzeit.setText("");
        lblDauer.setText("");
        cbHD.setSelected(false);
        cbSubtitle.setSelected(false);
        lblGeo.setText("");
        lblAbo.setText("");
        Platform.runLater(() -> {
            hyperlink.setTooltip(null);
            hyperlink.setDisable(true);
        });
    }

    private void setSenderIcon(final JLabel control) {
        final Optional<ImageIcon> optIcon = senderIconCache.get(currentFilm.getSender(), true);
        optIcon.ifPresent(icon -> {
            control.setText("");
            control.setIcon(icon);
        });
    }

    private void updateTextFields() {
        if (currentFilm == null) {
            clearControls();
        } else {
            setSenderIcon(lblSender);

            lblThema.setText(currentFilm.getThema());
            lblTitel.setText(currentFilm.getTitle());

            lblDescription.setText(currentFilm.getDescription());
            lblDescription.setCaretPosition(0);

            lblSize.setText(currentFilm.getSize());
            lblDatum.setText(currentFilm.getSendeDatum());
            lblUhrzeit.setText(currentFilm.getSendeZeit());
            lblDauer.setText(currentFilm.getDauer());

            cbHD.setSelected(currentFilm.isHD());
            cbSubtitle.setSelected(currentFilm.hasSubtitle());

            lblGeo.setText(currentFilm.getGeo().orElse(""));
            lblAbo.setText(currentFilm.getAboName());

            Platform.runLater(() -> {
                hyperlink.setTooltip(new Tooltip(currentFilm.getWebsiteLink()));
                hyperlink.setDisable(false);
            });
        }
    }

    /**
     * This will set the display to the new film data.
     *
     * @param film the film data
     */
    public void updateCurrentFilm(DatenFilm film) {
        currentFilm = film;
        if (isVisible())
            updateTextFields();
        pack();
    }

    private void buildLayout() {
        final Container contentPane = getContentPane();
        contentPane.setLayout(new MigLayout(
                "hidemode 3",
                //columns
                "[fill]" +
                        "[fill]",
                //rows
                "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[]" +
                        "[fill]"));

        JLabel label = new JLabel();
        label.setText("Sender:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,0));

        lblSender = new JLabel();
        contentPane.add(lblSender, new CC().cell(1,0).maxWidth("250"));

        label = new JLabel();
        label.setText("Thema:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label,new CC().cell(0,1));

        lblThema = new MultilineLabel();
        contentPane.add(lblThema, new CC().cell(1,1));

        label = new JLabel();
        label.setText("Titel:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,2));

        lblTitel = new MultilineLabel();
        contentPane.add(lblTitel, new CC().cell(1,2).maxWidth("250"));

        label = new JLabel();
        label.setText("Datum:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,3));

        lblDatum = new JLabel();
        contentPane.add(lblDatum, new CC().cell(1,3).maxWidth("250"));

        label = new JLabel();
        label.setText("Uhrzeit:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,4));

        lblUhrzeit = new JLabel();
        contentPane.add(lblUhrzeit, new CC().cell(1,4).maxWidth("250"));

        label = new JLabel();
        label.setText("Dauer:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,5));

        lblDauer = new JLabel();
        contentPane.add(lblDauer, new CC().cell(1,5).maxWidth("250"));

        label = new JLabel();
        label.setText("Größe (MB):");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,6));

        lblSize = new JLabel();
        contentPane.add(lblSize, new CC().cell(1,6).maxWidth("250"));

        label = new JLabel();
        label.setText("HD:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,7));

        cbHD = new JCheckBox();
        cbHD.setEnabled(false);
        contentPane.add(cbHD, new CC().cell(1,7).maxWidth("250"));

        label = new JLabel();
        label.setText("Untertitel:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,8));

        cbSubtitle = new JCheckBox();
        cbSubtitle.setEnabled(false);
        contentPane.add(cbSubtitle, new CC().cell(1,8).maxWidth("250"));

        label = new JLabel();
        label.setText("Geo:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,9));

        lblGeo = new JLabel();
        contentPane.add(lblGeo, new CC().cell(1,9).maxWidth("250"));

        label = new JLabel();
        label.setText("Abo:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,10));

        lblAbo = new JLabel();
        contentPane.add(lblAbo, new CC().cell(1,10).maxWidth("250"));

        label = new JLabel();
        label.setText("Website:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,11));

        fxPanel = new JFXPanel();
        contentPane.add(fxPanel, new CC().cell(1,11).maxWidth("250"));

        Platform.runLater(() -> {
            ContextMenu contextMenu = new ContextMenu();
            MenuItem mi = new MenuItem(COPY_URL_TEXT);
            mi.setOnAction(e -> SwingUtilities.invokeLater(() -> GuiFunktionen.copyToClipboard(currentFilm.getWebsiteLink())));
            contextMenu.getItems().add(mi);

            hyperlink = new Hyperlink("Hier klicken");
            hyperlink.setContextMenu(contextMenu);
            hyperlink.setUnderline(true);
            hyperlink.setBackground(new Background(new BackgroundFill(Color.rgb(236, 236, 236), CornerRadii.EMPTY, Insets.EMPTY)));
            hyperlink.setOnAction(e -> SwingUtilities.invokeLater(() -> {
                if (currentFilm != null) {
                    try {
                        UrlHyperlinkAction.openURL(null, currentFilm.getWebsiteLink());
                    } catch (URISyntaxException ex) {
                        ex.printStackTrace();
                    }
                }
            }));
            fxPanel.setScene(new Scene(hyperlink));
        });
        label = new JLabel();
        label.setText("Beschreibung:");
        label.setHorizontalAlignment(SwingConstants.RIGHT);
        contentPane.add(label, new CC().cell(0,12));

        final JScrollPane scrollPane = new JScrollPane();
        lblDescription = new JTextArea();
        lblDescription.setLineWrap(true);
        lblDescription.setEditable(false);
        lblDescription.setWrapStyleWord(true);
        lblDescription.setFont(UIManager.getFont("Label.font"));
        scrollPane.setViewportView(lblDescription);

        contentPane.add(scrollPane, "cell 0 13 2 1,growy,width 250:320:400,height 100:100");
        pack();
    }
}

File: src/main/java/mediathek/gui/filmInformation/InfoDialog.kt
package mediathek.gui.filmInformation

import com.formdev.flatlaf.util.ScaledImageIcon
import com.jidesoft.swing.MultilineLabel
import mediathek.daten.DatenFilm
import mediathek.gui.actions.UrlHyperlinkAction
import mediathek.mainwindow.MediathekGui
import mediathek.tool.ApplicationConfiguration
import mediathek.tool.GuiFunktionen
import mediathek.tool.SwingErrorDialog
import mediathek.tool.sender_icon_cache.MVSenderIconCache
import net.miginfocom.layout.AC
import net.miginfocom.layout.CC
import net.miginfocom.layout.LC
import net.miginfocom.swing.MigLayout
import org.apache.commons.configuration2.sync.LockMode
import org.jdesktop.swingx.JXHyperlink
import java.awt.Desktop
import java.awt.Dimension
import java.awt.Point
import java.awt.Window
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.net.URI
import java.net.URISyntaxException
import java.util.concurrent.atomic.AtomicBoolean
import javax.swing.*


class InfoDialog(parent: Window?) : JDialog(parent) {
    private val config = ApplicationConfiguration.getConfiguration()
    private var currentFilm: DatenFilm? = null
    private val lblSender = JLabel()
    private val lblThema = MultilineLabel()
    private val lblTitel = MultilineLabel()
    private val lblDate = JLabel()
    private val lblUhrzeit = JLabel()
    private val lblDuration = JLabel()
    private val lblSize = JLabel()
    private val lblGeo = JLabel()
    private val cbHq = SwingDisabledCheckBox()
    private val cbSubtitle = SwingDisabledCheckBox()
    private val lblAbo = JLabel()
    private val hyperlink = SwingHyperlink()
    private val descScrollPane = JScrollPane()
    private val lblDescription = JTextArea()
    private val isPacking = AtomicBoolean(false)


    internal class SwingDisabledCheckBox : JCheckBox() {
        init {
            isEnabled = false
        }
    }
    /**
     * Restore window position from config settings.
     */
    private fun restoreLocation() {
        config.lock(LockMode.READ)
        try {
            val newLocation = Point()
            newLocation.x = config.getInt(ApplicationConfiguration.FilmInfoDialog.FILM_INFO_LOCATION_X)
            newLocation.y = config.getInt(ApplicationConfiguration.FilmInfoDialog.FILM_INFO_LOCATION_Y)
            location = newLocation
        } catch (ignored: NoSuchElementException) {
        } finally {
            config.unlock(LockMode.READ)
        }
    }

    override fun pack() {
        isPacking.set(true)
        super.pack()
        restoreLocation()
        isPacking.set(false)
    }

    /**
     * Save window position to config.
     */
    private fun saveLocation() {
        //prevent strange OpenJDK bug on Linux where getLocationOnScreen will fail if not visible...
        if (!isVisible) return
        config.lock(LockMode.WRITE)
        try {
            val location = locationOnScreen
            config.setProperty(ApplicationConfiguration.FilmInfoDialog.FILM_INFO_LOCATION_X, location.x)
            config.setProperty(ApplicationConfiguration.FilmInfoDialog.FILM_INFO_LOCATION_Y, location.y)
        } finally {
            config.unlock(LockMode.WRITE)
        }
    }

    fun showInfo() {
        updateTextFields()
        if (!isVisible) isVisible = true
    }

    private fun clearControls() {
        lblThema.text = ""
        lblSender.text = ""
        lblSender.icon = null
        lblTitel.text = ""
        lblDate.text = ""
        lblUhrzeit.text = ""
        lblDuration.text = ""
        lblSender.text = ""
        lblGeo.text = ""
        cbHq.isSelected = false
        cbSubtitle.isSelected = false
        lblAbo.text = ""
        hyperlink.toolTipText = ""
        hyperlink.isEnabled = false
        lblDescription.text = ""
        SwingUtilities.invokeLater { pack() }
    }

    companion object {
        private val DEFAULT_SENDER_DIMENSION = Dimension(64,64)
    }
    private fun updateTextFields() {
        if (currentFilm == null) {
            clearControls()
        } else {
            MVSenderIconCache[currentFilm!!.sender].ifPresent { icon: ImageIcon? ->
                lblSender.text = ""
                val imageDim = Dimension(icon!!.iconWidth, icon.iconHeight)
                val destDim = GuiFunktionen.calculateFittedDimension(imageDim, DEFAULT_SENDER_DIMENSION)
                lblSender.icon = ScaledImageIcon(icon, destDim.width, destDim.height)
            }
            lblThema.text = currentFilm!!.thema
            lblTitel.text = currentFilm!!.title
            lblDate.text = currentFilm!!.sendeDatum
            lblUhrzeit.text = currentFilm!!.sendeZeit
            lblDuration.text = currentFilm!!.filmLengthAsString
            lblSize.text = currentFilm!!.fileSize.toString()
            lblGeo.text = currentFilm!!.geo.orElse("")
            cbHq.isSelected = currentFilm!!.isHighQuality
            cbSubtitle.isSelected = currentFilm!!.hasSubtitle()
            lblAbo.text = currentFilm!!.abo?.name
            hyperlink.isEnabled = true
            hyperlink.toolTipText = currentFilm!!.websiteUrl
            hyperlink.isClicked = false

            val desc = currentFilm!!.description.trim { it <= ' ' }
            lblDescription.text = desc
            SwingUtilities.invokeLater {
                descScrollPane.verticalScrollBar.value = 0
                pack()
            }
        }
    }

    /**
     * This will set the display to the new film data.
     *
     * @param film the film data
     */
    fun updateCurrentFilm(film: DatenFilm?) {
        currentFilm = film
        if (isVisible) {
            updateTextFields()
            //pack()
        }
    }

    internal class SwingHyperlink : JXHyperlink() {
        private fun openUrl(url : String) {
            try {
                UrlHyperlinkAction.openURL(null, url)
            } catch (ex: URISyntaxException) {
                ex.printStackTrace()
            }
        }

        init {
            this.text = "Link zur Webseite"

            addActionListener {
                if (toolTipText.isNotEmpty()) {
                    if (Desktop.isDesktopSupported()) {
                        val d = Desktop.getDesktop()
                        if (d.isSupported(Desktop.Action.BROWSE)) {
                            try {
                                d.browse(URI(toolTipText))
                            }
                            catch (ex: Exception) {
                                SwingErrorDialog.showExceptionMessage(
                                    MediathekGui.ui(),
                                    "Es trat ein Fehler beim Öffnen des Links auf.\nSollte dies häufiger auftreten kontaktieren Sie bitte das Entwicklerteam.",
                                    ex)
                            }
                        } else {
                            openUrl(toolTipText)
                        }
                    }
                    else {
                        openUrl(toolTipText)
                    }

                }
            }
        }
    }

    private fun buildLayout() {
        layout = MigLayout(
            LC().insets("5").hideMode(3).debug(),
            // columns
            AC()
                .fill().gap()
                .size("250!").fill(),
            // rows
            AC()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap()
                .gap())

        add(JLabel("Sender:"), CC().cell(0, 0))
        add(lblSender, CC().cell(1, 0))
        add(JLabel("Thema:"), CC().cell(0, 1))
        lblThema.lineWrap = true
        add(lblThema, CC().cell(1, 1).growY())
        add(JLabel("Titel:"), CC().cell(0, 2))
        lblTitel.lineWrap = true
        add(lblTitel, CC().cell(1, 2).growY())

        add(JLabel("Datum:"), CC().cell(0, 3))
        add(lblDate, CC().cell(1, 3))
        add(JLabel("Uhrzeit:"), CC().cell(0, 4))
        add(lblUhrzeit, CC().cell(1, 4))
        add(JLabel("Dauer:"), CC().cell(0, 5))
        add(lblDuration, CC().cell(1, 5))
        add(JLabel("Größe (MB):"), CC().cell(0, 6))
        add(lblSize, CC().cell(1, 6))
        add(JLabel("HQ:"), CC().cell(0, 7))
        add(cbHq, CC().cell(1, 7))
        add(JLabel("Untertitel:"), CC().cell(0, 8))
        add(cbSubtitle, CC().cell(1, 8))
        add(JLabel("Geo:"), CC().cell(0, 9))
        add(lblGeo, CC().cell(1, 9))
        add(JLabel("Abo:"), CC().cell(0, 10))
        add(lblAbo, CC().cell(1, 10))
        add(hyperlink, CC().cell(0, 11).spanX(2))
        add(JLabel("Beschreibung:"), CC().cell(0, 12))
        lblDescription.isEditable = false
        lblDescription.lineWrap = true
        lblDescription.wrapStyleWord = true
        descScrollPane.setViewportView(lblDescription)
        add(descScrollPane, CC().cell(0, 13).spanX(2).growY().growX().height("90!"))
    }

    init {
        type = Type.UTILITY
        title = "Filminformation"
        isResizable = false
        //hardcode size as linux hates pack()
        //setSize(350, 520)
//        minimumSize = Dimension(350,450)
        defaultCloseOperation = DISPOSE_ON_CLOSE
        buildLayout()
        pack()
        updateTextFields()
        restoreLocation()
        val wasVisible = config.getBoolean(ApplicationConfiguration.FilmInfoDialog.FILM_INFO_VISIBLE, false)
        if (wasVisible) {
            isVisible = true
        }
        addWindowListener(object : WindowAdapter() {
            override fun windowOpened(e: WindowEvent) {
                config.setProperty(ApplicationConfiguration.FilmInfoDialog.FILM_INFO_VISIBLE, true)
            }

            override fun windowClosed(e: WindowEvent) {
                config.setProperty(ApplicationConfiguration.FilmInfoDialog.FILM_INFO_VISIBLE, false)
            }
        })

        //addFilmlistLoadListener();
        addComponentListener(object : ComponentAdapter() {
            override fun componentMoved(e: ComponentEvent) {
                if (isVisible) {
                    if (!isPacking.get())
                        saveLocation()
                }
            }
        })
    }
}